src/checksum.cc

```diff
@@ -0,0 +1,226 @@
+  TRITONSERVER_Error* err_;
+};
+
+#define THROW_IF_TRITON_ERROR(X)                                    \
+  do {                                                              \
+    TRITONSERVER_Error* tie_err__ = (X);                            \
+    if (tie_err__ != nullptr) {                                     \
+      throw triton::repoagent::checksum::ErrorException(tie_err__); \
+    }                                                               \
+  } while (false)
+
+
+#define THROW_TRITON_ERROR(CODE, MSG)                                 \
+  do {                                                                \
+    TRITONSERVER_Error* tie_err__ = TRITONSERVER_ErrorNew(CODE, MSG); \
+    throw triton::repoagent::checksum::ErrorException(tie_err__);     \
+  } while (false)
+
+
+#define RETURN_IF_ERROR(X)               \
+  do {                                   \
+    TRITONSERVER_Error* rie_err__ = (X); \
+    if (rie_err__ != nullptr) {          \
+      return rie_err__;                  \
+    }                                    \
+  } while (false)
+
+class CheckSum {
+ public:
+  virtual std::string AlgorithmTypeString() = 0;
+  virtual std::string GenerateHash(const std::string& message) = 0;
+};
+
+class MD5Sum : public CheckSum {
+ public:
+  std::string AlgorithmTypeString() override { return "MD5"; }
+
+  std::string GenerateHash(const std::string& message) override
+  {
+    MD5((unsigned char*)message.data(), message.size(), result_);
+    unsigned char low_mask = 0b00001111;
+    unsigned char high_mask = 0b11110000;
+    std::stringstream stream;
+    for (size_t idx = 0; idx < MD5_DIGEST_LENGTH; idx++) {
+      stream << std::hex << (size_t)((result_[idx] & high_mask) >> 4)
+             << std::hex << (size_t)(result_[idx] & low_mask);
+    }
+    return stream.str();
+  }
+
+ private:
+  unsigned char result_[MD5_DIGEST_LENGTH];
+};
+
+std::pair<RelativeFilePath, HashString>
+ParseFileAndChecksum(const std::string& formatted_checksum_output)
+{
+  auto pos = formatted_checksum_output.find(" ");
+  if (pos == std::string::npos) {
+    THROW_TRITON_ERROR(
+        TRITONSERVER_ERROR_INVALID_ARG,
+        (std::string("failed to parse checksum output: ") +
+         formatted_checksum_output)
+            .c_str());
+  }
+  return {formatted_checksum_output.substr(pos + 2),
+          formatted_checksum_output.substr(0, pos)};
+}
+
+const std::string
+ReadFile(const std::string& model_dir, const std::string& relative_file_path)
+{
+  std::ifstream in(
+      model_dir + "/" + relative_file_path, std::ios::in | std::ios::binary);
+  if (!in) {
+    THROW_TRITON_ERROR(
+        TRITONSERVER_ERROR_INVALID_ARG,
+        (std::string("failed to open text file for read ") +
+         relative_file_path + ": " + strerror(errno))
+            .c_str());
+  }
+
+  std::string res;
+  in.seekg(0, std::ios::end);
+  res.resize(in.tellg());
+  in.seekg(0, std::ios::beg);
+  in.read(&res[0], res.size());
```

Comment:
It would be better to somehow stream through the file instead of reading the whole file in. I think this is ok for an example but you should add a comment saying that this isn't the most efficient way to do it.